src/sessions/PageManager.ts

```diff
@@ -91,31 +162,143 @@ export class PageManager {
+                wrapper: this.fetchWrapper
+            }
+        ];
+    }
+
+    private sendWrapper = (): ((original: Send) => Send) => {
+        const self = this;
+
+        return (original: Send): Send => {
+            return function (this: XMLHttpRequest): void {
+                self.recordXhr(this);
+                this.addEventListener('loadend', self.endTracking);
+                return original.apply(this, arguments);
+            };
+        };
+    };
+
+    private recordXhr(item: XMLHttpRequest) {
+        const page = this.page;
+        if (page && page.isLoaded !== null && page.isLoaded === false) {
+            page.ongoingActivity.add(item);
+        } else {
+            this.requestBuffer.add(item);
+        }
+    }
+
+    private removeXhr(item: XMLHttpRequest, currTime: number) {
+        const page = this.page;
+        if (page && page.ongoingActivity.has(item)) {
+            page.ongoingActivity.delete(item);
+            page.latestEndTime = Math.max(page.latestEndTime, currTime);
+        } else if (this.requestBuffer.has(item)) {
+            this.requestBuffer.delete(item);
+        }
+    }
+
+    /**
+     * Removes the current event from either requestBuffer or ongoingActivity set.
+     * @param event
+     */
+    private endTracking = (e: Event) => {
+        const currTime = Date.now();
+        const xhr: XMLHttpRequest = e.target as XMLHttpRequest;
+        xhr.removeEventListener('loadend', this.endTracking);
+        this.removeXhr(xhr, currTime);
+    };
+
+    private fetch = (
+        original: Fetch,
+        thisArg: Fetch,
+        argsArray: IArguments,
+        input: RequestInfo,
+        init?: RequestInit
+    ): Promise<Response> => {
+        const self = this;
+        return original
+            .apply(thisArg, argsArray)
+            .then((response: Response) => {
+                self.updateFetchCounter();
+            })
+            .catch((error) => {
+                self.updateFetchCounter();
+            });
+    };
+
+    /**
+     * Increment the fetch counter in PageManager when fetch is beginning
+     */
+    private fetchWrapper = (): ((
+        original: (input: RequestInfo, init?: RequestInit) => Promise<Response>
+    ) => (input: RequestInfo, init?: RequestInit) => Promise<Response>) => {
+        const self = this;
+        return (original: Fetch): Fetch => {
+            return function (
+                this: Fetch,
+                input: RequestInfo,
+                init?: RequestInit
+            ): Promise<Response> {
+                self.fetchCounter += 1;
+                return self.fetch(original, this, arguments, input, init);
+            };
+        };
+    };
+
+    private updateFetchCounter() {
+        const page = this.page;
+        // prevent NPE
+        if (page && !page.isLoaded) {
+            page.latestEndTime = Math.max(page.latestEndTime, Date.now());
+        }
+        if (this.fetchCounter > 0) {
```

Comment:
question: In what situation would fetchCounter be `0`?

Reply:
Ideally, fetchCounter should never be 0, since for every fetch, whether successful or erroneous, is decrementing. This was mainly for just a safety check just in case the counter goes lower than 0. What do you think?

Reply:
To go below zero, either (1) the counter fails to increment, or (2) `updateFetchCounter` gets executed multiple times for the same `fetch` call. I don't see a path where either of these could happen.